pkg/sdkserver/sdkserver.go

```diff
@@ -449,6 +449,17 @@ func (s *SDKServer) gameServer() (*agonesv1.GameServer, error) {
 	return gs, nil
 }
 
+// patchGameServer is a helper function to create and apply a patch update, so the changes in
+// gsCopy are applied to the original gs.
+func (s *SDKServer) patchGameServer(ctx context.Context, gs, gsCopy *agonesv1.GameServer) (*agonesv1.GameServer, error) {
+	patch, err := gs.Patch(gsCopy)
```

Comment:
Does this `gs.Patch(...)` do the optimistic locking?

Reply:
It does not. Client-go's .Patch (which the `s.gameServerGetter.GameServers(...).Patch(...)` is generated from) doesn't have the option as part of the PatchOptions, so it needs to be passed as part of the patch. There is an MergeFromWithOptimisticLock option in the https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client#MergeFromWithOptions (Strategic Merge doesn't support CRDs)

```
patch := client.MergeFromWithOptions(gsCopy, client.MergeFromWithOptimisticLock{})
patchData, err := patch.Data(gs)
```

This is the type `"application/merge-patch+json"`, which I believe we decided not use, and to instead use type `"application/json-patch+json"`. I also haven't tested this yet in conjunction with Client-go's .Patch. The only examples I've seen use sigs.k8s's client to patch.

The other option would be to include the Resource Version as a "test" operation `jsonpatch.JsonPatchOperation{{Operation: "test", Path: "/metadata/resourceVersion", Value: gs.ObjectMeta.ResourceVersion}}` as a part of the list of patch operations in `gs.Patch(...)`. Per https://jsonpatch.com/ "Tests that the specified value is set in the document. If the test fails, then the patch as a whole should not apply." This also means that if the ResourceVersion is not set the Patch would fail. Given that the Game Server should always have the ResourceVesion set I'm leaning towards this option.

Do you have any thoughts on which one we should use, or if there's another better option?

Reply:
The issue actually discussed this - I think we can use the same way controller-manager does: https://github.com/kubernetes-sigs/controller-runtime/blob/c25fe2f5ca0f9e10d571f85da61c3d210d6345e7/pkg/client/patch.go#L110-L114, which just ensures the resourceVersion makes it into the patch, AIUI.

Reply:
> The issue actually discussed this - I think we can use the same way controller-manager does: https://github.com/kubernetes-sigs/controller-runtime/blob/c25fe2f5ca0f9e10d571f85da61c3d210d6345e7/pkg/client/patch.go#L110-L114, which just ensures the resourceVersion makes it into the patch, AIUI.

I'm relatively sure @igooch is right on this one - that example is a mergePatch, whereas we are using a JSONPatch - so it's different.

Reply:
> extend gs.Patch(gs *GameServer, ...options) to have a set of options, with the only one right now being WithOptimisticLocking (like the runtime-controller).

I think we can start without it being optional. We're moving from optimistically-locked Update, I'd rather just start with optimistically-locked Patch and gradually claw back.

Reply:
Huh, I swore we used that function in a bunch of places, but apparently not!

Yep - no need for options!

Reply:
And @igooch you're totally right, great work here! Looks good, reviewing the whole again now.